File: net.akehurst.transform.binary/src/test/java/net/akehurst/transform/binary/test_Transformer.java
/**
 * Copyright (C) 2015 Dr. David H. Akehurst (http://dr.david.h.akehurst.net)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package net.akehurst.transform.binary;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import org.junit.Assert;
import org.junit.Test;
import org.junit.runner.RunWith;

import com.tngtech.java.junit.dataprovider.DataProvider;
import com.tngtech.java.junit.dataprovider.DataProviderRunner;
import com.tngtech.java.junit.dataprovider.UseDataProvider;

import net.akehurst.transform.binary.api.TransformException;

@RunWith(DataProviderRunner.class)
public class test_Transformer {

    @DataProvider
    public static Object[][] data_updateAllLeft2Right() {
        //@formatter:off
        return new Object[][] {
            { null, null, TransformException.class },
            { null, new Object[] {}, TransformException.class },
            { new Object[] {}, null, TransformException.class },
            { new Object[] {}, new Object[] {}, new Object[] {} },
            {
                new Object[] {},
                new Object[] { new BElement(1), new BElement(2), new BElement(3) },
                new Object[] {}
            },
            { new Object[] { new AElement(1), new AElement(2), new AElement(3) },
              new Object[] {},
              new Object[] { new BElement(1), new BElement(2), new BElement(3) }
            },
            { new Object[] { new AElement(1) },
              new Object[] { new BElement(1) },
              new Object[] { new BElement(1) }
            },
            { new Object[] { new AElement(1) },
              new Object[] { new BElement(2) },
              new Object[] { new BElement(1) }
            },
            { new Object[] { new AElement(2) },
              new Object[] { new BElement(1) },
              new Object[] { new BElement(2) }
            },
            { new Object[] { new AElement(1), new AElement(2) },
              new Object[] { new BElement(1) },
              new Object[] { new BElement(1), new BElement(2) }
            },
            { new Object[] { new AElement(1) },
              new Object[] { new BElement(1), new BElement(2) },
              new Object[] { new BElement(1) }
            },
            { new Object[] { new AElement(1), new AElement(2) },
              new Object[] { new BElement(2) },
              new Object[] { new BElement(1), new BElement(2) }
            },
            { new Object[] { new AElement(2) },
              new Object[] { new BElement(1), new BElement(2) },
              new Object[] { new BElement(2) }
            }
        };
        //@formatter:on
    }

    @Test
    @UseDataProvider("data_updateAllLeft2Right")
    public void test_updateAllLeft2Right(final Object[] leftArr, final Object[] rightArr, final Object result) {
        try {
            final List<AElement> leftList = null == leftArr ? null : new ArrayList(Arrays.asList(leftArr));
            final List<BElement> rightList = null == rightArr ? null : new ArrayList(Arrays.asList(rightArr));
            final ExampleTransformer transformer = new ExampleTransformer();
            transformer.updateAllLeft2Right(AElement2BElement.class, leftList, rightList);

            final List<BElement> resultList = null == result ? null : new ArrayList(Arrays.asList((Object[]) result));
            Assert.assertEquals(resultList, rightList);

        } catch (final Exception e) {
            if (e.getClass().equals(result)) {
                // ok
            } else {
                e.printStackTrace();
                Assert.fail(e.getMessage());
            }
        }
    }

    @DataProvider
    public static Object[][] data_updateAllRight2Left() {
        return new Object[][] {
            //
            { null, null, TransformException.class },
            //
            { null, new Object[] {}, TransformException.class },
            //
            { new Object[] {}, null, TransformException.class }, /**/ { new Object[] {}, new Object[] {}, /**/ new Object[] {} },
            //
            { new Object[] {}, /**/ new Object[] { new BElement(1), new BElement(2), new BElement(3) },
                /**/ new Object[] { new AElement(1), new AElement(2), new AElement(3) } },
            //
            { new Object[] { new AElement(1), new AElement(2), new AElement(3) }, new Object[] {}, new Object[] {} },
            //
            { new Object[] { new AElement(1) }, /**/ new Object[] { new BElement(1) }, /**/ new Object[] { new AElement(1) } },
            //
            { new Object[] { new AElement(1) }, /**/ new Object[] { new BElement(2) }, /**/ new Object[] { new AElement(2) } },
            //
            { new Object[] { new AElement(2) }, /**/ new Object[] { new BElement(1) }, /**/ new Object[] { new AElement(1) } },
            //
            { new Object[] { new AElement(1), new AElement(2) }, /**/ new Object[] { new BElement(1) }, /**/ new Object[] { new AElement(1) } },
            //
            { new Object[] { new AElement(1) }, /**/ new Object[] { new BElement(1), new BElement(2) },
                /**/ new Object[] { new AElement(1), new AElement(2) } },
            //
            { new Object[] { new AElement(1), new AElement(2) }, /**/ new Object[] { new BElement(2) }, /**/ new Object[] { new AElement(2) } },
            //
            { new Object[] { new AElement(2) }, /**/ new Object[] { new BElement(1), new BElement(2) }, /**/ new Object[] { new AElement(1), new AElement(2) } }
            //
        };
    }

    @Test
    @UseDataProvider("data_updateAllRight2Left")
    public void test_updateAllRight2Left(final Object[] leftArr, final Object[] rightArr, final Object result) {
        try {
            final List<AElement> leftList = null == leftArr ? null : new ArrayList(Arrays.asList(leftArr));
            final List<BElement> rightList = null == rightArr ? null : new ArrayList(Arrays.asList(rightArr));
            final ExampleTransformer transformer = new ExampleTransformer();
            transformer.updateAllRight2Left(AElement2BElement.class, leftList, rightList);

            final List<BElement> resultList = null == result ? null : new ArrayList(Arrays.asList((Object[]) result));
            Assert.assertEquals(resultList, leftList);

        } catch (final Exception e) {
            if (e.getClass().equals(result)) {
                // ok
            } else {
                e.printStackTrace();
                Assert.fail(e.getMessage());
            }
        }
    }
}

File: net.akehurst.transform/transform-binary/src/jvm8Test/kotlin/binary/test_Transformer.kt
/**
 * Copyright (C) 2019 Dr. David H. Akehurst (http://dr.david.h.akehurst.net)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package net.akehurst.transform.binary.test

import org.junit.runner.RunWith
import org.junit.runners.Parameterized
import org.junit.runners.Parameterized.Parameter
import org.junit.runners.Parameterized.Parameters
import net.akehurst.transform.binary.api.TransformException
import net.akehurst.transform.binary.test.model.AElement
import net.akehurst.transform.binary.test.model.BElement
import net.akehurst.transform.binary.test.transformer.ABTestTransformer
import net.akehurst.transform.binary.test.transformer.AElement2BElement
import java.io.BufferedReader
import java.io.InputStreamReader
import kotlin.test.Test
import kotlin.test.assertEquals

@RunWith(Parameterized::class)
class test_Transformer(val data:Data) {

    companion object {

        data class Data(
                val left:Any,
                val right:Any
        )

        @JvmStatic
        @Parameters(name = "{0}")
        fun data(): List<Data> {
            return listOf(
                    Data( AElement(1), BElement(1) )

            )
        }

    }

    @Test
    fun isAMatch() {
        val sut = ABTestTransformer()

        val actual = sut.isAMatch(AElement2BElement::class, this.data.left as AElement, this.data.right as BElement)

        assertEquals(true, actual)
    }

    @Test
    fun transformLeft2Right() {
        val sut = ABTestTransformer()

        val actual = sut.transformLeft2Right(AElement2BElement::class, this.data.left as AElement)

        assertEquals(this.data.right as BElement, actual)
    }


    @Test
    fun transformRight2Left() {
        val sut = ABTestTransformer()

        val actual = sut.transformRight2Left(AElement2BElement::class, this.data.right as BElement)

        assertEquals(this.data.left as AElement, actual)
    }
}
